lib/Sema/TypeCheckMacros.cpp

```diff
@@ -42,22 +42,44 @@ extern "C" ptrdiff_t swift_ASTGen_evaluateMacro(
     const char **evaluatedSource, ptrdiff_t *evaluatedSourceLength);
 
 extern "C" void
-swift_ASTGen_getMacroTypeSignature(void *macro, const char **genericSignature,
-                                   ptrdiff_t *genericSignatureLength);
+swift_ASTGen_getMacroTypeSignature(void *macro,
+                                   const char **evaluationContextPtr,
+                                   ptrdiff_t *evaluationContextLengthPtr);
+
+static NullTerminatedStringRef
+getPluginMacroTypeSignature(CompilerPlugin *plugin, ASTContext &ctx) {
+  auto genSig = plugin->invokeGenericSignature();
+  auto typeSig = plugin->invokeTypeSignature();
+  std::string source;
+  llvm::raw_string_ostream out(source);
+  out << "struct __MacroEvaluationContext" << (genSig ? *genSig : "") << " {\n"
+      << "  typealias SignatureType = " << typeSig << "\n"
+      << "}";
+  auto len = source.length();
+  auto *buffer = (char *)malloc(len + 1);
+  memcpy(buffer, source.data(), len + 1);
+  return {buffer, len};
+}
 
 StructDecl *MacroContextRequest::evaluate(Evaluator &evaluator,
                                           std::string macroName,
                                           ModuleDecl *mod) const {
 #if SWIFT_SWIFT_PARSER
   auto &ctx = mod->getASTContext();
-  auto *macro = swift_ASTGen_lookupMacro(macroName.c_str());
-  if (!macro)
+  auto *builtinMacro = swift_ASTGen_lookupMacro(macroName.c_str());
+  NullTerminatedStringRef evaluatedSource;
+  if (builtinMacro) {
+    const char *evaluatedSourcePtr;
+    ptrdiff_t evaluatedSourceLength;
+    swift_ASTGen_getMacroTypeSignature(builtinMacro, &evaluatedSourcePtr,
```

Comment:
It's interesting how we have similar code on both sides of the C++ divide. I want to think about this more.